import grpc
import json
import uuid
import time

import order_pb2
import order_pb2_grpc
import auth_pb2
import auth_pb2_grpc

def test_saga_scenario():
    print("üîß –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Saga –ø–∞—Ç—Ç–µ—Ä–Ω–∞...")
    
    # 1. –°–æ–∑–¥–∞–µ–º –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è (–∏–ª–∏ –∏—Å–ø–æ–ª—å–∑—É–µ–º —Å—É—â–µ—Å—Ç–≤—É—é—â–µ–≥–æ)
    auth_channel = grpc.insecure_channel('localhost:50051')
    auth_stub = auth_pb2_grpc.AuthServiceStub(auth_channel)
    
    # –†–µ–≥–∏—Å—Ç—Ä–∏—Ä—É–µ–º —Ç–µ—Å—Ç–æ–≤–æ–≥–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è
    try:
        auth_response = auth_stub.Register(auth_pb2.RegisterRequest(
            email=f"test_{uuid.uuid4().hex[:8]}@example.com",
            password="test123",
            phone_number="+79991234567",
            role="CUSTOMER",
            bank_account="1234567890"
        ))
        user_id = auth_response.user_id
        print(f"‚úÖ –°–æ–∑–¥–∞–Ω –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å: {user_id}")
    except:
        # –ò—Å–ø–æ–ª—å–∑—É–µ–º —Å—É—â–µ—Å—Ç–≤—É—é—â–µ–≥–æ
        user_id = "sample_user_1"
        print(f"‚úÖ –ò—Å–ø–æ–ª—å–∑—É–µ–º —Å—É—â–µ—Å—Ç–≤—É—é—â–µ–≥–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è: {user_id}")
    
    # 2. –ü–æ–¥–∫–ª—é—á–∞–µ–º—Å—è –∫ Order Service
    order_channel = grpc.insecure_channel('localhost:50053')
    order_stub = order_pb2_grpc.OrderServiceStub(order_channel)
    
    # 3. –î–æ–±–∞–≤–ª—è–µ–º —Ç–æ–≤–∞—Ä –≤ –∫–æ—Ä–∑–∏–Ω—É
    print("üõí –î–æ–±–∞–≤–ª—è–µ–º —Ç–æ–≤–∞—Ä –≤ –∫–æ—Ä–∑–∏–Ω—É...")
    try:
        # –ò—Å–ø–æ–ª—å–∑—É–µ–º —Å—É—â–µ—Å—Ç–≤—É—é—â–∏–π —Ç–æ–≤–∞—Ä
        product_id = "prod_001"
        
        cart_response = order_stub.AddToCart(order_pb2.AddToCartRequest(
            user_id=user_id,
            product_id=product_id,
            quantity=1
        ))
        
        if cart_response.success:
            cart_product_id = cart_response.cart_product_id
            print(f"‚úÖ –¢–æ–≤–∞—Ä –¥–æ–±–∞–≤–ª–µ–Ω –≤ –∫–æ—Ä–∑–∏–Ω—É: {cart_product_id}")
        else:
            print("‚ùå –ù–µ —É–¥–∞–ª–æ—Å—å –¥–æ–±–∞–≤–∏—Ç—å —Ç–æ–≤–∞—Ä –≤ –∫–æ—Ä–∑–∏–Ω—É")
            return
    except Exception as e:
        print(f"‚ùå –û—à–∏–±–∫–∞ –¥–æ–±–∞–≤–ª–µ–Ω–∏—è –≤ –∫–æ—Ä–∑–∏–Ω—É: {str(e)}")
        return
    
    # 4. –ü–æ–∫—É–ø–∞–µ–º —Ç–æ–≤–∞—Ä (–∑–∞–ø—É—Å–∫–∞–µ–º Saga)
    print("üí∞ –ü–æ–∫—É–ø–∞–µ–º —Ç–æ–≤–∞—Ä (–∑–∞–ø—É—Å–∫ Saga)...")
    try:
        buy_response = order_stub.BuyFromCart(order_pb2.BuyFromCartRequest(
            user_id=user_id,
            cart_product_id=cart_product_id,
            bank_details="1234567890"
        ))
        
        if buy_response.success:
            order_id = buy_response.order_id
            print(f"‚úÖ –ó–∞–∫–∞–∑ —Å–æ–∑–¥–∞–Ω: {order_id}")
            print("üîÑ Saga –∑–∞–ø—É—â–µ–Ω–∞: Order ‚Üí RabbitMQ ‚Üí Payment ‚Üí Notification")
        else:
            print("‚ùå –ù–µ —É–¥–∞–ª–æ—Å—å —Å–æ–∑–¥–∞—Ç—å –∑–∞–∫–∞–∑")
            return
    except Exception as e:
        print(f"‚ùå –û—à–∏–±–∫–∞ —Å–æ–∑–¥–∞–Ω–∏—è –∑–∞–∫–∞–∑–∞: {str(e)}")
        return
    
    # 5. –ü—Ä–æ–≤–µ—Ä—è–µ–º —Å—Ç–∞—Ç—É—Å –∑–∞–∫–∞–∑–∞ —á–µ—Ä–µ–∑ –Ω–µ—Å–∫–æ–ª—å–∫–æ —Å–µ–∫—É–Ω–¥
    print("\n‚è≥ –ñ–¥–µ–º –æ–±—Ä–∞–±–æ—Ç–∫–∏ –ø–ª–∞—Ç–µ–∂–∞ (5 —Å–µ–∫—É–Ω–¥)...")
    time.sleep(5)
    
    try:
        order_response = order_stub.GetOrder(order_pb2.GetOrderRequest(
            order_id=order_id
        ))
        
        print(f"\nüì¶ –°—Ç–∞—Ç—É—Å –∑–∞–∫–∞–∑–∞ {order_id}: {order_response.status}")
        
        if order_response.status == 'CONFIRMED':
            print("‚úÖ Saga —É—Å–ø–µ—à–Ω–æ –∑–∞–≤–µ—Ä—à–µ–Ω–∞!")
            print("   - –ó–∞–∫–∞–∑ —Å–æ–∑–¥–∞–Ω")
            print("   - –ü–ª–∞—Ç–µ–∂ —É—Å–ø–µ—à–Ω–æ –æ–±—Ä–∞–±–æ—Ç–∞–Ω")
            print("   - –£–≤–µ–¥–æ–º–ª–µ–Ω–∏—è –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω—ã")
            print("   - –¢–æ–≤–∞—Ä —Å–ø–∏—Å–∞–Ω —Å–æ —Å–∫–ª–∞–¥–∞")
        elif order_response.status == 'FAILED':
            print("üîÑ Saga –≤—ã–ø–æ–ª–Ω–µ–Ω–∞ —Å –∫–æ–º–ø–µ–Ω—Å–∞—Ü–∏–µ–π!")
            print("   - –ó–∞–∫–∞–∑ —Å–æ–∑–¥–∞–Ω")
            print("   - –ü–ª–∞—Ç–µ–∂ –Ω–µ –ø—Ä–æ—à–µ–ª")
            print("   - –¢–æ–≤–∞—Ä –≤–æ–∑–≤—Ä–∞—â–µ–Ω –≤ –∫–æ—Ä–∑–∏–Ω—É")
            print("   - –£–≤–µ–¥–æ–º–ª–µ–Ω–∏—è –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω—ã")
        else:
            print("‚è≥ Saga –≤—Å–µ –µ—â–µ –æ–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç—Å—è...")
            
    except Exception as e:
        print(f"‚ùå –û—à–∏–±–∫–∞ –ø—Ä–æ–≤–µ—Ä–∫–∏ —Å—Ç–∞—Ç—É—Å–∞: {str(e)}")
    
    # 6. –ü—Ä–æ–≤–µ—Ä—è–µ–º RabbitMQ
    print("\nüìä –ü—Ä–æ–≤–µ—Ä—å—Ç–µ RabbitMQ:")
    print("   http://localhost:15672")
    print("   –õ–æ–≥–∏–Ω: guest, –ü–∞—Ä–æ–ª—å: guest")
    print("   –û—á–µ—Ä–µ–¥—å 'payment_queue' –¥–æ–ª–∂–Ω–∞ —Å–æ–¥–µ—Ä–∂–∞—Ç—å —Å–æ–æ–±—â–µ–Ω–∏—è")

if __name__ == '__main__':
    test_saga_scenario()